src/main/kotlin/gg/ingot/iron/model/ModelReader.kt

```diff
@@ -0,0 +1,135 @@
+package gg.ingot.iron.model
+
+import gg.ingot.iron.Iron
+import gg.ingot.iron.annotations.Model
+import gg.ingot.iron.annotations.retrieveDeserializer
+import gg.ingot.iron.annotations.retrieveSerializer
+import gg.ingot.iron.representation.EntityField
+import gg.ingot.iron.representation.EntityModel
+import gg.ingot.iron.sql.params.ColumnJsonField
+import gg.ingot.iron.sql.params.ColumnSerializedField
+import gg.ingot.iron.strategies.NamingStrategy
+import org.jetbrains.annotations.Nullable
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.reflect.KClass
+import kotlin.reflect.full.declaredMemberProperties
+import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.jvm.javaField
+
+/**
+ * Transforms a model to an entity representation which holds information about the class model, allowing
+ * for the model to be easily built. This class interfaces with reflection to build the entity.
+ * @author santio
+ * @since 2.0
+ */
+class ModelReader(
+    private val iron: Iron
+) {
+
+    /**
+     * Transforms a class into an entity model, which holds information about the reflected class.
+     * @param clazz The class to transform into an entity model.
+     * @return The entity model representation of the class.
+     */
+    fun read(clazz: Class<*>): EntityModel {
+        return models.getOrPut(clazz) {
+            generateModel(clazz)
+        }
+    }
+
+    /**
+     * Transforms a class into an entity model, which holds information about the reflected class.
+     * @param clazz The kotlin class to transform into an entity model.
+     * @return The entity model representation of the class.
+     */
+    fun read(clazz: KClass<*>): EntityModel {
+        return read(clazz.java)
+    }
+
+    private fun generateModel(clazz: Class<*>): EntityModel {
+        val entityFields = mutableListOf<EntityField>()
+        val modelAnnotation = clazz.annotations.find { it is Model } as Model?
+
+        val fields = if (clazz.kotlin.isData) {
+            val constructor = clazz.kotlin.primaryConstructor
+            var properties = clazz.kotlin.declaredMemberProperties
+                .filter { it.javaField != null }
+
+            if (constructor != null) {
+                properties = properties.sortedBy {
+                    constructor.parameters.indexOfFirst { param -> param.name == it.name }
+                }
+            }
+
+            properties.map { ModelField(iron, it, it.javaField!!) }
+        } else {
+            clazz.declaredFields
+                .map { ModelField(iron, null, it) }
+        }.filter { !it.isIgnored() }
+
+        // Check if we have multiple primary keys
+        val primaryKey = fields.filter { it.isPrimaryKey() }
+        if (primaryKey.size > 1) {
```

Comment:
why can't we just use `check` function here?

Reply:
Stale comment